### ISSUE TITLE
Should `format_collect` be warn by default?

### ISSUE THREAD

**Issue by smoelius** [rust-lang/rust-clippy#11434]
### Description

[`format_collect`](https://github.com/rust-lang/rust-clippy/pull/11116) suggests to replace `format!` with `write!`.

A downside of this suggestion is that it replaces a infallible function with a fallible one (i.e., `write!` returns a `Result`, whereas `format!` does not).

[`format_push_string`](https://rust-lang.github.io/rust-clippy/master/index.html#/format_push_string) has [a similar downside](https://github.com/rust-lang/rust-clippy/issues/9077), and it was [moved to restriction](https://github.com/rust-lang/rust-clippy/pull/9161) for this reason.

Should the same standard apply to `format_collect`? Or is something about its situation different?

cc: @y21, @Centri3

**Comment by y21** [comment]
Yeah, we were discussing the category a bit on the PR and centri3 suggested pedantic I think. I wasn't aware of that lint. I do kinda agree now that warn by default might be a bit too strong and maybe this should be in restriction, seeing that `format_push_string` is also and the arguments given.
It's a bit unfortunate that `write!` on a string, [an operation that can't fail](https://doc.rust-lang.org/stable/src/alloc/string.rs.html#2879), returns a `Result`, thus requiring an unnecessary `let _ = ` or `.unwrap()`, but it makes sense given that the macro goes through the generic `std::fmt::Write` machinery.

**Comment by y21** [comment]
Right, so I think `format_collect` and `format_push_string` should probably be in the same category, since they're basically checking for the same kind of pattern but in different places. So there's two options, then:

1) Move `format_collect` to restriction and leave `format_push_string` where it is
2) Move `format_collect` *and* `format_push_string` to pedantic

I personally think 2) is better (I haven't found a strong argument as to why it should be restriction specifically), but I created a poll on zulip for this:
https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60format_collect.60.20category
